subprojects/core/src/integTest/groovy/org/gradle/api/tasks/TaskTimeoutIntegrationTest.groovy

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks
+
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.workers.IsolationMode
+import spock.lang.Unroll
+
+class TaskTimeoutIntegrationTest extends AbstractIntegrationSpec {
+
+    @Unroll
+    def "timeout stops long running method call"() {
+        given:
+        buildFile << """
+            task block() {
+                doLast {
+                    Thread.sleep(60000)
+                }
+                timeoutAfter 500, 'MILLISECONDS'
+            }
+            """
+
+        expect:
+        fails "block"
```

Comment:
We could check the duration of the task. The build duration itself is highly variable.

Reply:
I tried, but these are variable as well. E.g. we might not be able to actually stop after 500ms, because we might be in the middle of a non-interruptible operation like forking a process.

Reply:
How about a relatively lenient assertion, e.g. taskDuration < 30s?

Reply:
I'm gonna experiment some more.